### PR TITLE
Compile and test MongooseIM with Erlang/OTP 20

### PR DESCRIPTION
**Pull request by michalwski**
With this PR it's possible to compile and test MongooseIM with Erlang/OTP from 18 to 20

Proposed changes include:
* updated dependencies to be compatible with Erlang/OTP from 18 to 20
   * fix to backwards incompatible changes in `worker_pool`
* update test to be able to run with Erlang/OTP 20:
   * updated escalus as in esl/escalus/150
   * updated meck to 0.8.6
   * change remaining `crypt:rand_bytes` to `strong_rand_bytes`

TODO:
* [x] fix (or even remove) SASL-DIGEST authentication (there is one test case failing in `login_SUITE`)
* [x] fix dialyzer errors
* [x]  configure travis jobs so that most of them run with Erlang/OTP 19.3, dialyzer and one other job runs with Erlang/OTP 20 and another job for Erlang/OTP 18.3
* [x] remove possible hacks for Erlang/OTP 17.x
* [x] update documentation so that it's clear MongooseIM will work with Erlang/OTP form 18 to 20
* [x] try testing MongooseIM with FIPS mode enabled
* [x] update documentation regarding FIPS mode (doc/developers-guide/OpenSSL-and-FIPS.md)
* [x] investigate decreased coverage

### CODE REVIEW

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 1]
```diff
@@ -494,18 +494,18 @@ restore_mnesia(Path) ->
     case ejabberd_admin:restore(Path) of
         {atomic, _} ->
             {ok, ""};
-        {error, Reason} ->
-            String = io_lib:format("Can't restore backup from ~p at node ~p: ~p",
-                                   [filename:absname(Path), node(), Reason]),
-            {cannot_restore, String};
         {aborted, {no_exists, Table}} ->
             String = io_lib:format("Can't restore backup from ~p at node ~p: Table ~p does not exist.",
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_admin.erl]: According to **Elvis**:
> Line 498 is too long:             String = io_lib:format("Can't restore backup from ~p at node ~p: Table ~p does not exist.",.

File: doc/developers-guide/OpenSSL-and-FIPS.md [hunk 2]
```diff
@@ -10,24 +10,25 @@ Currently known incompatible features are:
...
 ```
-KERL_CONFIGURE_OPTIONS="--enable-fips --with-ssl=/home/vagrant/openssl" \
-./kerl build git https://github.com/michalwski/otp.git fips-17.4.1 17.4.1-fips
+KERL_CONFIGURE_OPTIONS="--enable-fips" ./kerl build 20.0 20.0-fips
 ```
 
 ### Building MongooseIM with a custom OpenSSL
 
-Before running `./rebar3 compile` or `make rel` please export the CFLAGS and LDFLAGS env vars pointing to a FIPS compliant OpenSSL, f.e.
+If you want to use a different OpenSSL than the default one, before running `./rebar3 compile` or `make rel` please export the CFLAGS and LDFLAGS env vars pointing to a FIPS compliant OpenSSL, f.e.
```
Comment (goddammit) [inline doc/developers-guide/OpenSSL-and-FIPS.md]: If you want to use a custom OpenSSL, please export the CFLAGS and LDFLAGS env vars pointing to a FIPS compliant OpenSSL before running `./rebar3 compile` or `make rel`.
F.e:

**Comment by DenysGonchar** [comment]
coverage decrease analysis, analysed coverage decrease between [base master commit](https://coveralls.io/builds/13298816) and [latest otp-20 commit](https://coveralls.io/builds/13373536)

| file_name | cov1* | cov2* | d_cov* | d_l1* | d_l2* |
|-|-|-|-|-|-|
| mongoose_fips.erl | 75.0 | 41.176 | -33.823 | 9 | 13 |
| ELDAPv3.erl | 22.712 | 14.708 | -8.004 | 690 | 690 |
| mongoose_client_api_sse.erl | 81.818 | 75.757 | -6.060 | 2 | 0 |
| eldap_filter_yecc.erl | 20.232 | 14.262 | -5.970 | 180 | 180 |
| mod_aws_sns.erl | 91.860 | 87.209 | -4.651 | 4 | 0 |
| mod_mam_odbc_prefs.erl | 98.901 | 95.604 | -3.296 | 3 | 0 |
| mod_mam_utils.erl | 88.086 | 85.920 | -2.166 | 6 | 0 |
| mod_muc.erl | 68.508 | 67.127 | -1.381 | 5 | 0 |
| mod_mam.erl | 89.802 | 89.473 | -0.328 | 1 | 0 |
| mod_muc_log.erl | 77.948 | 77.692 | -0.256 | 1 | 0 |
| node_flat.erl | 74.006 | 73.993 | -0.012 | -1 | -4 |


* cov1  - coverage for base master commit
  cov2  - coverage for the last otp-20 commit
  d_cov - change of the coverage
  d_l1  - change of covered lines No
  d_l2  - change of lines No in the file

**Comment by michalwski** [comment]
@DenysGonchar this is impressive what you've prepared! I'm wondering why there are these differences in coverage reports. Most of the modules didn't change in this PR... Maybe we can live with the way coverage is reported now?

**Comment by fenek** [comment]
@michalwski 

`ELDAPv3` and `eldap_filter_yecc` are generated by OTP so it seems they have changed the algorithm to create more verbose files and we don't use these new parts. `mongoose_fips` has a `ifndef` flag that depends on FIPS support and it seems we don't have any specific tests for it.

@DenysGonchar 

What are these lines that are suddenly not covered in the remaining files?

**Comment by michalwski** [comment]
Yes, coverage change in `mongoose_fips` is expected. Regarding the LDAP-related modules, will we still need them when the old PR #1216 is merged?

**Comment by DenysGonchar** [comment]
guys let me finish...

On Sep 29, 2017 15:45, "Michał Piotrowski" <notifications@github.com> wrote:

Yes, coverage change in mongoose_fips is expected. Regarding the
LDAP-related modules, will we still need them when the old PR #1216
<https://github.com/esl/MongooseIM/pull/1216> is merged?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/esl/MongooseIM/pull/1430#issuecomment-333130328>, or mute
the thread

**Comment by DenysGonchar** [comment]
some comments on the changed files:

**ELDAPv3.erl** and **eldap_filter_yecc.erl** are generated and there's no sense to analyse them, also we might want to exclude that modules from coverage analysis at all.

**mongoose_fips.erl** - as mentioned above, otp20 has support of fips_mode so rebar adds the corresponding compilation flag, from here we have new lines. so far we don't have tests for fips_mode so decrease of coverage is expected.

**mongoose_client_api_sse.erl** - uncovered lines are:
   * dummy handle_info/2 clause
   * and handle_error/3 function

**mod_aws_sns.erl** - uncovered lines are:
   *  catch block in try_publish/5 function. more over for base commit catch block is not fully covered, it means that calc_backoff_time/2 function fails & generates exception. we should take a closer look on this!!!
   *   calc_backoff_time/2 function (seems that exception is generated at [***line 294***](https://coveralls.io/builds/13298816/source?filename=apps%2Fejabberd%2Fsrc%2Fmod_aws_sns.erl#L294))

**mod_mam_odbc_prefs** - uncovered lines are:
   * catch block in run_transaction_or_retry_on_deadlock/3 function

**mod_mam_utils.elr** - uncovered lines are:
   * is_mam_result_message/1 function and 2 more internal functions called from this one (maybe_get_result_namespace/1 and is_mam_namespace/1)
   * error clause of error_on_sql_error/3 function

**mod_muc.erl** - uncovered lines are:
   * processing function passed to lists:foreach/2 inside load_permanent_rooms/6, RoomsToLoad is an empty list
   * default case clause in iq_disco_items/4 (line 805, clause just returns false)

**mod_mam.erl** - uncovered lines are:
   * sm_filter_offline_message/4

**mod_muc_log.erl** - difference in coverage is weekdays clauses in get_dateweek/2 function, this is explained by the simple fact that tests were running in different weekdays. and for base commit 
 it seems that one of the travis jobs failed and was restarted on the next day :)

**node_flat.erl**  -  difference in coverage caused by removed code in init/1, removed code is case block that has 2 clauses, one returns - ok, another one wasn't covered in base commit. so nothing to worry about.


## none of these looks critical to me

**Comment by michalwski** [comment]
Thanks @DenysGonchar the analysation you did is impressive. Looks like the decrease in coverage will not be harmful for us. Merging this PR now has bigger value then fighting with the coverage. The sad thing is that we'll have less 70% coverage again. There is a quick fix to this situation - merging #1216 :)

**Comment by DenysGonchar** [comment]
i'm not 100% sure that it brings us back to 70+% but we can try exclude those generated files from coverage analysis.

**Pull request merged**